src/pluggy/hooks.py

```diff
@@ -360,3 +361,142 @@ def __init__(self, namespace, name, opts):
         self.opts = opts
         self.argnames = ["__multicall__"] + list(self.argnames)
         self.warn_on_impl = opts.get("warn_on_impl")
+
+
+def _raise_wrapfail(wrap_controller, msg):
+    co = wrap_controller.gi_code
+    raise RuntimeError(
+        "wrap_controller at %r %s:%d %s"
+        % (co.co_name, co.co_filename, co.co_firstlineno, msg)
+    )
+
+
+class HookCallError(Exception):
+    """ Hook was called wrongly. """
+
+
+def _wrapped_call(wrap_controller, func):
+    """ Wrap calling to a function with a generator which needs to yield
+    exactly once.  The yield point will trigger calling the wrapped function
+    and return its ``_Result`` to the yield point.  The generator then needs
+    to finish (raise StopIteration) in order for the wrapped call to complete.
+    """
+    try:
+        next(wrap_controller)  # first yield
+    except StopIteration:
+        _raise_wrapfail(wrap_controller, "did not yield")
+    call_outcome = _Result.from_call(func)
+    try:
+        wrap_controller.send(call_outcome)
+        _raise_wrapfail(wrap_controller, "has second yield")
+    except StopIteration:
+        pass
+    return call_outcome.get_result()
+
+
+class _LegacyMultiCall(object):
```

Comment:
This reminds me, we can drop `_LegacyMultiCall` by now (https://github.com/pytest-dev/pluggy/issues/59).

Reply:
There is an open PR for this #147. Would certainly simplify things to drop it, if it's now possible.

Reply:
Yes that should land before we start moving big chunks of code around. Let's simplify first then worry about a reorg.